### PR TITLE
Rework frontend API spec

### PR DESCRIPTION
**Pull request by OxygenCobalt**
Rework the frontend API spec to:
- Make it easier to use from the frontend, mostly by separating waypoints and stops, separating outages and alerts, and defining typescript interfaces for requests/responses
- Explicitly use JSON instead of any ad-hoc format
- Define some new endpoints due to new stakeholder requirements discussed earlier

Other non-frontend points are currently disregarded.

### CODE REVIEW

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:47]: change to typescript for highlighting reasons

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md]: misspelled backend

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:83]: specify `/alert/stop/{stop_id}` and `/alert/route/{route_id}`

Comment (grantlemons) [inline backend/backend-schema-proposal.md:83]: why do this with a separate endpoint and not when retrieving routes/stops?

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:104]: Request body

File: backend/backend-schema-proposal.md
Comment (shanecranor) [inline backend/backend-schema-proposal.md:65]: why are routes not connected to their stops?

File: backend/backend-schema-proposal.md
Comment (shanecranor) [inline backend/backend-schema-proposal.md:72]: I'm thinking that stops should maybe be a property of route?

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:132]: seconds are good enough here, I was using millis for other timestamps as a way to invalidate past requests, but that level of precision is unneeded here

this could also be an rfc 3339 date if that is easier to deserialize

File: backend/backend-schema-proposal.md
Comment (shanecranor) [reply backend/backend-schema-proposal.md:72]: or Waypoint can have boolean isStop

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:43]: maybe add that it will show an alert on the admin dashboard that there is a significant amount of clock drift

**Review by grantlemons — CHANGES_REQUESTED** [review]
will look more closely when I have access to a computer

File: backend/backend-schema-proposal.md
Comment (shanecranor) [inline backend/backend-schema-proposal.md]: everything else is timestampMillis, should this also be?

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:83]: Good point. We don't want to underfetch, but merging it into `/routes` makes that even more overloaded. I think I'm going to merge it and rename it to `/dashboard`.

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md]: Resolved in 3bf7c73.

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:65]: Two reasons:
1. Multiple routes can share one stop. So, we can avoid fetching redundant data if we separate them.
2. By separating them, the frontend can draw the path of the routes, and then overlay stop icons on stop of them. If they were combined, we would have to sift through the waypoint data to identify stops, which is business logic that we don't want on the frontend.

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:132]: Fixed in 579fd46.

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:72]: See https://github.com/OreCart/OreCart-App/pull/33/files#r1349551715

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:43]: I don't fully get what you mean.

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:83]: Why exactly?

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:55]:
```suggestion
```typescript
interface Location {
    vanId: number,
    routeId: number,
    location: Coordinate,
    timestampMillis: number
}
```
```

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:1]: go ahead and clarify that hardware -> backend UNIX timestamps are in millis

Lines 9 & 22

Comment (grantlemons) [inline backend/backend-schema-proposal.md:1]: go ahead and add a timestamp to line 33 and clarify that it is a UNIX timestamp in millis

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:59]: something about this seems like bad practice

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:96]: I kinda wish we were authenticating now, is there any way we can restrict the amount of ADA requests in a short time from a single user?

I know we can do something on the backend, but this seems like a major pain point if a malicious actor discovers this.

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:110]: I think `/admin/outages` makes more sense for this and the POST one

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:137]:
```suggestion
- Stops reporting the location of all vans registered to a route or stop
```

File: backend/backend-schema-proposal.md
Comment (grantlemons) [inline backend/backend-schema-proposal.md:1]: let's add an admin endpoint for mapping vans to routes

this would let us disable vans (by assigning no routes) as well

**Review by grantlemons — CHANGES_REQUESTED** [review]

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:59]: Well, we don't want to underfetch, and I'm not comfortable with calling it `/routes` since it contains information completely unrelated to routes. Any ideas?

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:96]: I'd imagine FastAPI has some form of rate limiting built in.

File: backend/backend-schema-proposal.md
Comment (OxygenCobalt) [reply backend/backend-schema-proposal.md:1]: I don't think we need milliseconds for the ADA request. Can we just do seconds?

**Review by grantlemons — APPROVED** [review]
LGTM

**Review by ezrichards — APPROVED** [review]
Nice job! Please wait for a review from either @shanecranor or @tbwrigh to merge though.

File: backend/backend-schema-proposal.md
Comment (shanecranor) [inline backend/backend-schema-proposal.md:69]: should be alert?: Alert

**Review by shanecranor — COMMENTED** [review]
we can merge after fixing :)

**Review by shanecranor — APPROVED** [review]

**Pull request merged**
